lisp/bazel-build.el

```diff
@@ -25,38 +25,109 @@
+(require 'subr-x)
 
 (defun bazel-build (target)
   "Build a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel build ")))
+  (interactive (list (bazel-build--read-target "bazel build")))
   (bazel-build--run-bazel-command "build" target))
 
 (defun bazel-run (target)
   "Build and run a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel run ")))
+  (interactive (list (bazel-build--read-target "bazel run")))
   (bazel-build--run-bazel-command "run" target))
 
 (defun bazel-test (target)
   "Build and run a Bazel test TARGET."
-  (interactive (list (bazel-build--read-target "bazel test ")))
+  (interactive (list (bazel-build--read-target "bazel test")))
   (bazel-build--run-bazel-command "test" target))
 
+(defvar bazel-command "bazel"
+  "Bazel executable to run for subprocesses.")
+
+(defvar bazel-query-args '("query" "--noblock_for_lock")
+  "Bazel query subcommand and arguments.")
+
+(defvar bazel-run-commands-from-root nil
+  "A flag indicating to run commands from the workspace root.")
+
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
-  (compile
-   (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
-
-(defun bazel-build--read-target (prompt)
-  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
-  (let* ((file-name (buffer-file-name))
-         (workspace-root
-          (or (bazel-util-workspace-root file-name)
-              (user-error "Not in a Bazel workspace.  No WORKSPACE file found")))
-         (package-name
-          (or (bazel-util-package-name file-name workspace-root)
-              (user-error "Not in a Bazel package.  No BUILD file found")))
-         (initial-input (concat "//" package-name)))
-    (read-string prompt initial-input)))
+  (let ((command
+         (if bazel-run-commands-from-root
+             (let ((root-dir (or (bazel-util-workspace-root
+                                  (or (buffer-file-name)
+                                      (and dired-directory (directory-file-name dired-directory))))
+                                 (user-error "Could not find workspace."))))
+               (format "cd %s && %s %s %s" root-dir
+                       bazel-command command target))
+           (mapconcat #'shell-quote-argument (list bazel-command command target) " "))))
+    (compile command)))
+
+(defun bazel-build--process (&rest args)
+  "Run a bazel subcommand. Return stdout as string."
+  ;; Note: The current working directory of the subprocess is set to the current
+  ;; buffer's value of default-directory.
+  (let ((stderr-file (make-temp-file "bazel-stderr")))
+    (unwind-protect
+        (string-trim-right
+         (with-output-to-string
+           (let ((status
+                  (apply #'process-file bazel-command
+                         nil (list standard-output stderr-file) nil args)))
+             (if (/= status 0)
```

Comment:
This test is incorrect if the process exited with a signal, then `status` will be a string and `/=` will fail with a type error. Use `(not (eq status 0))` instead.